matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/aggregation/poll/DefaultPollAggregationProcessor.kt

```diff
@@ -0,0 +1,210 @@
+        val totalVotes = existingVotes.size
+        val newVotesSummary = existingVotes
+                .groupBy({ it.option }, { it.userId })
+                .mapValues {
+                    VoteSummary(
+                            total = it.value.size,
+                            percentage = if (totalVotes == 0 && it.value.isEmpty()) 0.0 else it.value.size.toDouble() / totalVotes
+                    )
+                }
+        val newWinnerVoteCount = newVotesSummary.maxOf { it.value.total }
+
+        if (isLocalEcho) {
+            aggregatedPollSummaryEntity.sourceLocalEchoEvents.add(event.eventId)
+        } else {
+            aggregatedPollSummaryEntity.sourceEvents.add(event.eventId)
+        }
+
+        val myVote = existingVotes.find { it.userId == session.myUserId }?.option
+
+        val newSumModel = PollSummaryContent(
+                myVote = myVote,
+                votes = existingVotes,
+                votesSummary = newVotesSummary,
+                totalVotes = totalVotes,
+                winnerVoteCount = newWinnerVoteCount
+        )
+        aggregatedPollSummaryEntity.aggregatedContent = ContentMapper.map(newSumModel.toContent())
+
+        return true
+    }
+
+    override fun handlePollEndEvent(session: Session, powerLevelsHelper: PowerLevelsHelper, realm: Realm, event: Event): Boolean {
+        val content = event.getClearContent()?.toModel<MessageEndPollContent>() ?: return false
+        val roomId = event.roomId ?: return false
+        val pollEventId = content.relatesTo?.eventId ?: return false
+        val pollOwnerId = getPollEvent(session, roomId, pollEventId)?.root?.senderId
+        val isPollOwner = pollOwnerId == event.senderId
+
+        if (!isPollOwner && !powerLevelsHelper.isUserAbleToRedact(event.senderId ?: "")) {
+            Timber.v("## Received poll.end event $pollEventId but user ${event.senderId} doesn't have enough power level in room $roomId")
+            return false
+        }
+
+        val annotationsSummaryEntity = getAnnotationsSummaryEntity(realm, roomId, pollEventId)
+        val aggregatedPollSummaryEntity = getAggregatedPollSummaryEntity(realm, annotationsSummaryEntity)
+
+        val txId = event.unsignedData?.transactionId
+        aggregatedPollSummaryEntity.closedTime = event.originServerTs
+
+        val isLocalEcho = LocalEcho.isLocalEchoId(event.eventId ?: "")
+        if (!isLocalEcho && aggregatedPollSummaryEntity.sourceLocalEchoEvents.contains(txId)) {
+            aggregatedPollSummaryEntity.sourceLocalEchoEvents.remove(txId)
+            aggregatedPollSummaryEntity.sourceEvents.add(event.eventId)
+        }
+
+        return true
+    }
+
+    private fun getPollEvent(session: Session, roomId: String, eventId: String): TimelineEvent? {
+        return session.roomService().getRoom(roomId)?.getTimelineEvent(eventId) ?: return null.also {
+            Timber.v("## POLL target poll event $eventId not found in room $roomId")
```

Comment:
Optional: If these are error logs we want to keep in prod, don't we want to increase the log level to info or warning? I think our verbose and debug logs are very overcrowded and it can make it harder to debug normal development tasks

Reply:
Removed. We don't need these logs since they are heavily tested.